src/Strategy/Archive.hs

```diff
@@ -0,0 +1,104 @@
+module Strategy.Archive
+  ( discover,
+  )
+where
+
+import qualified Codec.Archive.Tar as Tar
+import qualified Codec.Archive.Zip as Zip
+import qualified Codec.Compression.GZip as GZip
+import Control.Carrier.Diagnostics (FailureBundle (..), SomeDiagnostic (..))
+import Control.Effect.Exception (SomeException, mask, try)
+import Control.Effect.Finally
+import Control.Effect.Lift
+import Control.Effect.Output (output)
+import Control.Effect.TaskPool (forkTask)
+import qualified Data.ByteString.Lazy as BL
+import Data.Foldable (traverse_)
+import Data.List (isSuffixOf)
+import Discovery.Walk
+import Path
+import qualified Path.IO as PIO
+import Strategy.Archive.RPM (extractRpm)
+import Types
+import Prelude hiding (zip)
+
+-- Given a discover function to run over unarchived contents, discover projects
+discover :: HasDiscover sig m => (Path Abs Dir -> m ()) -> Path Abs Dir -> m ()
+discover go = walk $ \_ _ files -> do
+  let tars = filter (\file -> ".tar" `isSuffixOf` fileName file) files
+  traverse_ (\file -> forkArchiveDiscover $ withArchive tar file go) tars
+
+  let tarGzs = filter (\file -> ".tar.gz" `isSuffixOf` fileName file) files
+  traverse_ (\file -> forkArchiveDiscover $ withArchive tarGz file go) tarGzs
+
+  let zips = filter (\file -> ".zip" `isSuffixOf` fileName file) files
+  traverse_ (\file -> forkArchiveDiscover $ withArchive zip file go) zips
+
+  let jars = filter (\file -> ".jar" `isSuffixOf` fileName file) files
+  traverse_ (\file -> forkArchiveDiscover $ withArchive zip file go) jars
+
+  let rpms = filter (\file -> ".rpm" `isSuffixOf` fileName file) files
+  traverse_ (\file -> forkArchiveDiscover $ withArchive extractRpm file go) rpms
+
+  pure WalkContinue
+
+-- | Fork discovery of archive contents as a new task, catching exceptions
+forkArchiveDiscover :: HasDiscover sig m => m () -> m ()
+forkArchiveDiscover go = forkTask $ do
+  mask $ \restore -> do
+    (res :: Either SomeException a) <- try (restore go)
+    case res of
+      Left exc -> output (ProjectFailure ArchiveGroup "archive" (FailureBundle [] (SomeDiagnostic [] exc)))
+      Right () -> pure ()
+
+-- | Extract an archive to a temporary directory, and run the provided callback
+-- on the temporary directory. Archive contents are removed when the callback
+-- finishes.
+withArchive ::
+  (Has (Lift IO) sig m, Has Finally sig m) =>
+  -- | Archive extraction function
+  (Path Abs Dir -> Path Abs File -> m ()) ->
+  -- | Path to archive
+  Path Abs File ->
+  -- | Callback
+  (Path Abs Dir -> m ()) ->
+  m ()
+withArchive extract file go = do
+  tmpDir <- mkTempDir (fileName file)
```

Comment:
Does this not cause a collision between the filename and the newly created temp directory? I would think that we have to strip the file extension before creating the directory?

Reply:
The temp directory is created as a _system_-level temp directory (like in `/tmp`), and the name provided serves as a "template" or prefix that'll be used when creating the system temp directory.

If a folder with that name already exists, it'll arbitrarily append random garbage to the end of it